drivers/dma/dma_intel_adsp_gpdma.c

```diff
@@ -44,6 +44,7 @@ LOG_MODULE_REGISTER(dma_intel_adsp_gpdma);
 /* Device run time data */
 struct intel_adsp_gpdma_data {
 	struct dw_dma_dev_data dw_data;
+	bool enabled;
```

Comment:
Arguably this is a bug in the DMA client side if there is an unbalance between the start and stop.
The `dw_dma_start()` and `dw_dma_stop()` rely on correct user behavior as it is handing the pm_device_runtime get/put

Right, the bug might be in the dw_dma_start/stop() balancing?
is this fixes the issue?
```diff
diff --git a/drivers/dma/dma_dw_common.c b/drivers/dma/dma_dw_common.c
index bb000298385f..a893672d0922 100644
--- a/drivers/dma/dma_dw_common.c
+++ b/drivers/dma/dma_dw_common.c
@@ -446,7 +446,7 @@ int dw_dma_start(const struct device *dev, uint32_t channel)
 	/* validate channel */
 	if (channel >= DW_CHAN_COUNT) {
 		ret = -EINVAL;
-		goto out;
+		goto pm_get;
 	}
 
 	if (dw_dma_is_enabled(dev, channel)) {
@@ -535,6 +535,8 @@ int dw_dma_start(const struct device *dev, uint32_t channel)
 
 	/* enable the channel */
 	dw_write(dev_cfg->base, DW_DMA_CHAN_EN, DW_CHAN_UNMASK(channel));
+
+pm_get:
 	ret = pm_device_runtime_get(dev);
 
 out:
@@ -554,7 +556,7 @@ int dw_dma_stop(const struct device *dev, uint32_t channel)
 	}
 
 	if (!dw_dma_is_enabled(dev, channel) && chan_data->state != DW_DMA_SUSPENDED) {
-		goto out;
+		goto pm_put;
 	}
 
 #ifdef CONFIG_DMA_DW_HW_LLI
@@ -605,6 +607,8 @@ int dw_dma_stop(const struct device *dev, uint32_t channel)
 	}
 #endif
 	chan_data->state = DW_DMA_IDLE;
+
+pm_put:
 	ret = pm_device_runtime_put(dev);
 out:
 	return ret;
```

Reply:
@ujfalusi This alone won't fix as dw_dma_is_enabled() will still get called on a powered down device and lead to crash. 

But ack, the idea seems to be to inc/dec refcount when a DMA channel is actually enabled. Unfortunately, this won't work as only tracking state we have is the hw registers and we can't safely query them to see whether we shoudl decrease or not.

This seems a bit complicated indeed. It would seem safer that we get a reference whenever we talk to hardware, and that is not done here, which does not seem right. But to allow multiple start/stop, we need some clean way to account for usage. That will potentially mean powering up the hardware just to do a duplicate stop, but this would probably be still cleaner code to read and maintain.

Reply:
I see, yes, my diff is not correct for the reason that multiple start w/o stop is a possibility or even multiple start w/o a single stop (one shot block copies for example).

But what about using `pm_device_state_get()` and PM_DEVICE_STATE_ACTIVE instead of the local flag?
That would work even if `CONFIG_PM_DEVICE` is not enabled?

Reply:
and it might be possible to have it in the generic code in `drivers/dma/dma_dw_common.c`, not sure about that.